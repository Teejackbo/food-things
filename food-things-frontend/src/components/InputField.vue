<template>
  <div :class="{'inline-block': display }" class="form-field">
    <slot />
    <input
      :value="value"
      :style="`width: ${inputWidth};`"
      v-bind="attributes"
      :class="{ error, valid }"
      @input="input"
      v-on="listeners"
    >
  </div>
</template>

<script>
export default {
  name: 'InputField',

  inheritAttrs: false,

  props: {
    value: {
      type: [String, Boolean, Number],
      required: false,
    },
    inputWidth: {
      type: String,
      required: false,
    },
    display: {
      type: String,
      required: false,
    },
    error: {
      type: Boolean,
      required: false,
    },
    valid: {
      type: Boolean,
      required: false,
    },
  },

  methods: {
    input(event) {
      this.$emit('input', event.target.value)
    },
  },

  computed: {
    listeners() {
      // eslint-disable-next-line
      const { input, ...listeners } = this.$listeners
      return listeners
    },

    attributes() {
      return this.$attrs
    },
  },
}
</script>

<style lang="scss" scoped>
@import '../style/variables';

@keyframes shake {
  10%,
  90% {
    transform: translate3d(-1px, 0, 0);
  }

  20%,
  80% {
    transform: translate3d(2px, 0, 0);
  }

  30%,
  50%,
  70% {
    transform: translate3d(-4px, 0, 0);
  }

  40%,
  60% {
    transform: translate3d(4px, 0, 0);
  }
}

.form-field {
  margin: 0.5em 0;
}

input {
  background: $grey-light;
  border: 1px solid $grey;
  border-radius: 5px;
  box-shadow: 0px 1px 2px $shadow-lighter;
  color: $grey-darkest;
  display: block;
  font-family: $font-body;
  font-size: 15px;
  padding: 0.3em 0.7em;

  &.offset {
    padding-left: 2em;
  }

  &.valid {
    border: 1px solid $primary;
  }

  &.error {
    animation: shake 0.5s cubic-bezier(0.36, 0.07, 0.19, 0.97) both;
    border: 1px solid $red;
  }
}

.inline-block {
  display: inline-block;
  margin-right: 1em;
}
</style>
