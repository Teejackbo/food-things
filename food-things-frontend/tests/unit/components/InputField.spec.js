import { shallowMount } from '@vue/test-utils'
import InputField from '@/components/InputField.vue'

function getWrapper(label, inputWidth, display) {
  return shallowMount(InputField, {
    propsData: {
      label,
      inputWidth,
      display,
    },
    attrs: {
      placeholder: 'placeholder',
      type: 'text',
    },
  })
}

describe('<InputField />', () => {
  it('Should match snapshot.', () => {
    const wrapper = getWrapper()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('Should render a div with the class of form field.', () => {
    const wrapper = getWrapper()

    expect(wrapper.is('div')).toBe(true)
    expect(wrapper.classes()).toContain('form-field')
  })

  it('Should not render a label when no label prop is passed.', () => {
    const wrapper = getWrapper()

    expect(wrapper.find('label').exists()).toBe(false)
  })

  it('Should render a label with the label prop as text.', () => {
    const text = 'Test Label'
    const wrapper = getWrapper(text)
    const label = wrapper.find('label')

    expect(label.exists()).toBe(true)
    expect(label.text()).toBe(text)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('Should not apply a style tag without an inputWidth prop.', () => {
    const wrapper = getWrapper()
    const input = wrapper.find('input')

    expect(input.attributes().style).toBe(undefined)
  })

  it('Should apply the inputWidth prop to the input.', () => {
    const wrapper = getWrapper('Label', '5em')
    const input = wrapper.find('input')

    expect(input.attributes().style).toBe('width: 5em;')
    expect(wrapper.element).toMatchSnapshot()
  })

  it('Should not apply a class without a display prop.', () => {
    const wrapper = getWrapper()

    expect(wrapper.classes()).not.toContain('inline-block')
  })

  it('Should apply the correct class if display prop is inline-block.', () => {
    const wrapper = getWrapper('Label', '5em', 'inline-block')

    expect(wrapper.classes()).toContain('inline-block')
    expect(wrapper.element).toMatchSnapshot()
  })

  it('Should emit an input event when typed in.', () => {
    const wrapper = getWrapper()
    const input = wrapper.find('input')
    input.setValue('test')

    expect(wrapper.emitted()).toEqual({ input: [['test']] })
  })

  it('Should apply attributes to the input element.', () => {
    const wrapper = getWrapper()
    const input = wrapper.find('input')

    expect(input.attributes()).toEqual({
      placeholder: 'placeholder',
      type: 'text',
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
