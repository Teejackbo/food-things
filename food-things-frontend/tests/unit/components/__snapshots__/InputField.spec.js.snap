// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputField /> Should apply props correctly. 1`] = `
<div
  class="form-field inline-block"
>
   
  <input
    class=""
    placeholder="placeholder"
    style="width: 5em;"
    type="text"
  />
</div>
`;

exports[`<InputField /> Should match snapshot. 1`] = `
<div
  class="form-field"
>
   
  <input
    class=""
    placeholder="placeholder"
    type="text"
  />
</div>
`;
